### ISSUE TITLE
Duplicate Cookies with different Path's

### ISSUE THREAD

**Issue by marcelalburg** [panva/node-oidc-provider#403]
Hello,

i created a default serverless oidc-provider and when i make this curl i got cookies with a wrong path - /auth instead of /inteaction

```
curl -D - "localhost:3000/auth?client_id=test_client&response_type=code&scope=openid&redirect_url=https://example.com/xxx"
HTTP/1.1 302 Found
pragma: no-cache
cache-control: no-cache, no-store
location: /interaction/394e95c8-c590-448b-a0ea-61f8450ebf70
content-type: text/html; charset=utf-8
content-length: 129
date: Wed, 21 Nov 2018 08:40:07 GMT
connection: close
set-cookie: _grant=394e95c8-c590-448b-a0ea-61f8450ebf70; path=/auth/394e95c8-c590-448b-a0ea-61f8450ebf70; expires=Wed, 21 Nov 2018 08:50:07 GMT; httponly
set-cookie: _grant.sig=aeTTZdFCIxOt08zETYact6B8q78; path=/auth/394e95c8-c590-448b-a0ea-61f8450ebf70; expires=Wed, 21 Nov 2018 08:50:07 GMT; httponly

Redirecting to <a href="/interaction/394e95c8-c590-448b-a0ea-61f8450ebf70">/interaction/394e95c8-c590-448b-a0ea-61f8450ebf70
```

I debugged and enabled logging and see that someone is adding the same cookie with a different path:

`Serverless: [302] {"statusCode":302,"body":"Redirecting to <a href=\"/interaction/394e95c8-c590-448b-a0ea-61f8450ebf70\">/interaction/394e95c8-c590-448b-a0ea-61f8450ebf70</a>.","headers":{"pragma":"no-cache","cache-control":"no-cache, no-store","location":"/interaction/394e95c8-c590-448b-a0ea-61f8450ebf70","content-type":"text/html; charset=utf-8","content-length":"129","date":"Wed, 21 Nov 2018 08:40:07 GMT","connection":"close","Set-cookie":"_grant=394e95c8-c590-448b-a0ea-61f8450ebf70; path=/interaction/394e95c8-c590-448b-a0ea-61f8450ebf70; expires=Wed, 21 Nov 2018 08:50:07 GMT; httponly","sEt-cookie":"_grant.sig=aeTTZdFCIxOt08zETYact6B8q78; path=/interaction/394e95c8-c590-448b-a0ea-61f8450ebf70; expires=Wed, 21 Nov 2018 08:50:07 GMT; httponly","SEt-cookie":"_grant=394e95c8-c590-448b-a0ea-61f8450ebf70; path=/auth/394e95c8-c590-448b-a0ea-61f8450ebf70; expires=Wed, 21 Nov 2018 08:50:07 GMT; httponly","seT-cookie":"_grant.sig=aeTTZdFCIxOt08zETYact6B8q78; path=/auth/394e95c8-c590-448b-a0ea-61f8450ebf70; expires=Wed, 21 Nov 2018 08:50:07 GMT; httponly"},"isBase64Encoded":false}
`

I know it's not exact only your product, but maybe you've an idea. Inside the oidc-provider i just see that ./actions/authorization/interactions.js is setting the correct cookies.

Thanks
Marcel

**Comment by panva** [comment]
Hi @marcelalburg,

this is intended and it's not someone, it's the provider :) see the responsible code [here](https://github.com/panva/node-oidc-provider/blob/v5.4.2/lib/actions/authorization/interactions.js#L207-L215).

The reason is that i don't wish to just rely on the path id component on the resume, the interaction session and it's resume must be bound to the user-agent so that you can't start an interaction on one user-agent, and someone else intercepting your resume call, blocking it, and opening it in his user-agent. Makes sense?

TLDR; this is so that every step of the authorization is bound to the user-agent of the end-user.

**Comment by panva** [comment]
ad serverless, are you using lambda to host? care share the adapter (i'm guessing dynamo) with the rest here?

**Comment by marcelalburg** [comment]
Hello, thanks for the super fast answer. I checked and debug this lines of code and the path at this lines is correct. 

and i using 90% of the code from https://github.com/AKIRA-MIYAKE/serverless-oidc-provider

Marcel

**Comment by marcelalburg** [comment]
And i use lamba proxy offered from aws.

**Comment by panva** [comment]
I think the adapter used in that project is not atomic when it comes to keeping track of tokens bound to a single grant, under load [this](https://github.com/AKIRA-MIYAKE/node-oidc-provider-dynamodb-adapter/blob/master/lib/index.js#L110-L131) part of the code doesn't maintain the list atomically.

**Comment by marcelalburg** [comment]
Ohh, interesting. Didn't know that the adapter can influences the cookies. I'm don't use this lib and copy the most of them to a custom code: https://gist.github.com/marcelalburg/284d09e40d86db6cc91a44ac73d9f7ce

but this don't handle this part you mention in you comment.

**Comment by panva** [comment]
this has nothing to do with the cookies, i was just interested in the adapter in use.
